### ISSUE TITLE
Save Website as Image: Not saving whole Cruncyroll history page & Save as CBZ instead of JPG/PNG 

### ISSUE THREAD

**Issue by ipkpjersi** [igorlogius/igorlogius#57]
### Can you reproduce the issue in a 2nd new unmodified firefox profile with only the addon installed in its latest version  on the current stable firefox

yes - Please only select if you actually did!

### Steps to Reproduce

1. Install the Save Website as Image Firefox addon.
2. Go to Crunchyroll history page.
3. Click Save Website as Image.

### Expected Result

I expect it to save the entire page in either JPG or PNG format.

### Actual Result

It saves the page as a cbz file, and it saves only the first part of the page as expected, it saves the rest of the Crunchyroll watch history page as black pages.

You can see an example of this here:
![image](https://github.com/user-attachments/assets/65d08c72-1c77-4abc-944d-e41aa7d655d5)

I've attached the cbz file as a zip archive file so you can see what I'm talking about:
[2024-10-22_19-08-19 Recent Anime History _ Crunchyroll https___www_crunchyroll_com_history.zip](https://github.com/user-attachments/files/17467717/2024-10-22_19-08-19.Recent.Anime.History._.Crunchyroll.https___www_crunchyroll_com_history.zip)

You can see the issue happens in a brand new Firefox profile with only this addon installed:
![image](https://github.com/user-attachments/assets/83c57d63-e929-4ce5-ad99-328ba645c7d8)

(I'm using Ubuntu 22.04 with xfce with a WinXP theme, although it likely doesn't matter)

**Comment by igorlogius** [comment]
Hi, thanks for taking the time to open this issue.

My first guess would be that the page is made in a way to lazy-load elements (aka. only load what is necessary to fill the viewport) and only when you scroll closer to the end of the  content, will the pages load more. 

To verify this, i'd request that you try holding the "Page Down" button to scroll down and see if this allows you to load everything (after a while). 

If it does, you should be able to generate the non-blank output, when clicking on the extension
button then after everything has been loaded.  

Please be aware that for such large pages the output as a CBZ is expected because of browser internal size limitations in regards to the so called canvas API. 

If you need the output to be a single images, you'll currently have to extract the CBZ and combine the images manually. 

Sorry for the inconvenience. 
I do plan to fix this when i find a solution for this problem.

**Comment by ipkpjersi** [comment]
Yep, it seems to be exactly as you said, the content seems to be lazy-loaded.

Although, I did scroll down before attempting to take the picture at first, which seems to be why it made the CBZ, the image was just so large. When I load the page without scrolling down at all, it produces a single 1920x9007 PNG image:
![2024-10-22_20-37-23 Recent Anime History _ Crunchyroll https___www_crunchyroll_com_history](https://github.com/user-attachments/assets/88255abc-214d-4fba-a3f0-df85adbb0ed2)

This is such a cool extension and I really appreciate that it's open source, and I do hope that one day this issue can be fixed. :slightly_smiling_face:

**Comment by igorlogius** [comment]
Some additional information.

Currently every addon and even the firefox built-in screenshot utility is limited by the canvas size restriction. 

> // The max dimension for a canvas is 32,767 https://searchfox.org/mozilla-central/rev/f40d29a11f2eb4685256b59934e637012ea6fb78/gfx/cairo/cairo/src/cairo-image-surface.c#62.
> // The max number of pixels for a canvas is 472,907,776 pixels (i.e., 22,528 x 20,992) https://developer.mozilla.org/en-US/docs/Web/HTML/Element/canvas#maximum_canvas_size
> // We have to limit screenshots to these dimensions otherwise it will cause an error.

src. https://searchfox.org/mozilla-central/rev/2455e4d6388cbdeadb44b9633b971a65a98b504c/browser/components/screenshots/ScreenshotsUtils.sys.mjs#43-48

Sadly stiching images together is not as simple as concatenating their raw-bytes together and i  havent found a JS library that would allow allows to combine png or jpg images without the use of the canvas API.

**Comment by igorlogius** [comment]
update and note: 

creating any kind of complete image object in memory sofar has failed (for larger sites)  simply because of the built-in size limitations in the basic TypedArrays. 

The only remaining idea that might still work is to emply something like (Readable/Writable) Streams to segment and process the generated data, but since i havent yet worked with those, i need to read up on them some more. 

tldr: currently one idea left, will take some time to prototype something
